### ISSUE TITLE
Suggest `collect`ing a range instead of pushing `i` onto vector in a `for` loop

### ISSUE THREAD

**Issue by mycielski** [rust-lang/rust-clippy#8325]
### What it does

This lint finds `for` loops (with incrementing `i`) used only to populate vectors with the value of `i` on each execution of the loop. The lint suggest the better way of populating a vector, that is using `collect` on a range. Please look in the code example to see a demonstration. I also invite you to run the benchmarks yourself to see how much faster the `collect` method is; I have written some test and published them in [this gist](https://gist.github.com/mycielski/fd3ceebcda6018b6277bf3d41d0a8ae6).

I have encountered this issue in the wild. In someone's code a vector was being filled in a for loop with consecutive numbers. To my great surprie `clippy` did not find this worrisome. That's why I am submitting this issue.

### Lint Name

needless_for_loop

### Category

perf

### Advantage

- Massive performance increase in comparison to a for loop.
- More compact, idiomatic code.

### Drawbacks

Populating a vector with a `for` loop guarantees that the vector is mutable. Populating a vector with range `collect`ion does not make such a guarantee. However if the lint is implemented correctly then it should not care about mutability nor should it change mutability if `--fix` is on.

### Example

```rust
let mut v: Vec<i32> = Vec::new();
for i in 0..50000 {
    v.push(i);
}
```

Could be written as:

```rust
let mut v = (0..50000).collect::<Vec<i32>>();
```

**Comment by camsteffen** [comment]
This applies to any iterator, not just ranges. Can probably be part of `manual_extend` suggested at #8169.

**Comment by alice-i-cecile** [comment]
Some manual cleanup of this in https://github.com/bevyengine/bevy/pull/5775 :)
